File: Backend/controllers/authController.js
const User = require("../models/User");
const jwt = require("jsonwebtoken");
const bcrypt = require("bcryptjs");

const generateToken = (user) => {
  return jwt.sign(
    { id: user._id, email: user.email, role: user.role },
    process.env.JWT_SECRET || "devconnect_jwt_secret_key_2024_secure_token",
    { expiresIn: "24h" }
  );
};

const SUPER_ADMIN_EMAIL = "adhithyanshanmugam@gmail.com";

exports.signin = async (req, res) => {
  try {
    const { email, password } = req.body;

    const user = await User.findOne({ email: email?.toLowerCase().trim() });
    if (!user) {
      return res.status(401).json({ message: "Invalid email or password" });
    }

    // Block login if not allowed by admin
    if (!user.isAllowed) {
      return res.status(403).json({
        message: "Your account has not been activated by admin. Please contact the administrator."
      });
    }

    const isMatch = await bcrypt.compare(password, user.password || "");
    if (!isMatch) {
      return res.status(401).json({ message: "Invalid email or password" });
    }

    const token = generateToken(user);

    res.status(200).json({
      token,
      id: user._id,
      email: user.email,
      role: user.role,
    });
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
};

exports.googleSignin = async (req, res) => {
  try {
    const { email, fullName, googleId } = req.body;
    const normalizedEmail = email?.toLowerCase().trim();

    let user = await User.findOne({ email: normalizedEmail });

    if (user) {
      // Existing user — check if allowed
      if (!user.isAllowed) {
        return res.status(403).json({
          message: "Your account registration request is pending admin approval. Please wait for activation."
        });
      }

      // Update fullName if not already set
      if (!user.fullName && fullName) {
        user.fullName = fullName;
        await user.save();
      }
    } else {
      // New user via Google — only super admin is allowed immediately, normal users are created inactive
      const isAdmin = normalizedEmail === SUPER_ADMIN_EMAIL;
      
      user = new User({
        email: normalizedEmail,
        fullName: fullName || "",
        role: isAdmin ? "admin" : "user",
        isAllowed: isAdmin,
        password: await bcrypt.hash(Math.random().toString(36).slice(-10), 10),
      });
      await user.save();

      if (!isAdmin) {
        return res.status(403).json({
          message: "Your account registration request has been submitted to the admin. Please wait for activation."
        });
      }
    }

    const token = generateToken(user);

    res.status(200).json({
      token,
      id: user._id,
      email: user.email,
      role: user.role,
    });
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
};

exports.signup = async (req, res) => {
  try {
    const { email, password, fullName } = req.body;
    if (!email || !password) {
      return res.status(400).json({ message: "Email and password are required" });
    }

    const normalizedEmail = email.toLowerCase().trim();
    const existingUser = await User.findOne({ email: normalizedEmail });

    if (existingUser) {
      if (!existingUser.isAllowed) {
        return res.status(400).json({
          message: "An account activation request for this email has already been submitted and is pending admin approval."
        });
      } else {
        return res.status(400).json({
          message: "An account with this email is already registered and activated. Please log in."
        });
      }
    }

    const hashedPassword = await bcrypt.hash(password, 10);
    const user = new User({
      email: normalizedEmail,
      password: hashedPassword,
      fullName: fullName || "",
      role: "user",
      isAllowed: false,
    });

    await user.save();

    res.status(201).json({
      message: "Registration request submitted successfully. Please wait for admin approval before logging in."
    });
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
};

File: Backend-Java/src/main/java/com/orrs/controller/AuthController.java
package com.orrs.controller;

import com.orrs.model.User;
import com.orrs.repository.UserRepository;
import com.orrs.security.JwtTokenProvider;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.web.bind.annotation.*;

import java.util.Map;
import java.util.Optional;
import java.util.UUID;

@RestController
@RequestMapping("/auth")
public class AuthController {

    private static final String SUPER_ADMIN_EMAIL = "adhithyanshanmugam@gmail.com";

    @Autowired
    private UserRepository userRepository;

    @Autowired
    private PasswordEncoder passwordEncoder;

    @Autowired
    private JwtTokenProvider jwtTokenProvider;

    @PostMapping("/signin")
    public ResponseEntity<?> signin(@RequestBody SigninRequest request) {
        try {
            if (request.getEmail() == null) {
                return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(Map.of("message", "Email is required"));
            }

            String email = request.getEmail().toLowerCase().trim();
            Optional<User> userOpt = userRepository.findByEmail(email);

            if (userOpt.isEmpty()) {
                return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body(Map.of("message", "Invalid email or password"));
            }

            User user = userOpt.get();

            if (!user.isAllowed()) {
                return ResponseEntity.status(HttpStatus.FORBIDDEN).body(Map.of(
                        "message", "Your account has not been activated by admin. Please contact the administrator."
                ));
            }

            String rawPassword = request.getPassword() != null ? request.getPassword() : "";
            if (!passwordEncoder.matches(rawPassword, user.getPassword() != null ? user.getPassword() : "")) {
                return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body(Map.of("message", "Invalid email or password"));
            }

            String token = jwtTokenProvider.generateToken(user.getId(), user.getEmail(), user.getRole());

            return ResponseEntity.ok(new AuthResponse(token, user.getId(), user.getEmail(), user.getRole()));
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(Map.of("message", e.getMessage()));
        }
    }

    @PostMapping("/google-signin")
    public ResponseEntity<?> googleSignin(@RequestBody GoogleSigninRequest request) {
        try {
            if (request.getEmail() == null) {
                return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(Map.of("message", "Email is required"));
            }

            String email = request.getEmail().toLowerCase().trim();
            Optional<User> userOpt = userRepository.findByEmail(email);
            User user;

            if (userOpt.isPresent()) {
                user = userOpt.get();
                if (!user.isAllowed()) {
                    return ResponseEntity.status(HttpStatus.FORBIDDEN).body(Map.of(
                            "message", "Your account has not been activated by admin. Please contact the administrator."
                    ));
                }

                // Update fullName if not set
                if ((user.getFullName() == null || user.getFullName().isEmpty()) && request.getFullName() != null) {
                    user.setFullName(request.getFullName());
                    userRepository.save(user);
                }
            } else {
                // New user - check if super admin
                boolean isSuperAdmin = email.equalsIgnoreCase(SUPER_ADMIN_EMAIL);
                if (!isSuperAdmin) {
                    return ResponseEntity.status(HttpStatus.FORBIDDEN).body(Map.of(
                            "message", "Your account has not been activated by admin. Please contact the administrator."
                    ));
                }

                user = new User();
                user.setEmail(email);
                user.setFullName(request.getFullName() != null ? request.getFullName() : "");
                user.setRole("admin");
                user.setAllowed(true);

                // Set a secure random password for OAuth user
                String randomPassword = UUID.randomUUID().toString().substring(0, 10);
                user.setPassword(passwordEncoder.encode(randomPassword));

                userRepository.save(user);
            }

            String token = jwtTokenProvider.generateToken(user.getId(), user.getEmail(), user.getRole());

            return ResponseEntity.ok(new AuthResponse(token, user.getId(), user.getEmail(), user.getRole()));
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(Map.of("message", e.getMessage()));
        }
    }

    public static class SigninRequest {
        private String email;
        private String password;

        public String getEmail() {
            return email;
        }

        public void setEmail(String email) {
            this.email = email;
        }

        public String getPassword() {
            return password;
        }

        public void setPassword(String password) {
            this.password = password;
        }
    }

    public static class GoogleSigninRequest {
        private String email;
        private String fullName;
        private String googleId;

        public String getEmail() {
            return email;
        }

        public void setEmail(String email) {
            this.email = email;
        }

        public String getFullName() {
            return fullName;
        }

        public void setFullName(String fullName) {
            this.fullName = fullName;
        }

        public String getGoogleId() {
            return googleId;
        }

        public void setGoogleId(String googleId) {
            this.googleId = googleId;
        }
    }

    public static class AuthResponse {
        private String token;
        private String id;
        private String email;
        private String role;

        public AuthResponse(String token, String id, String email, String role) {
            this.token = token;
            this.id = id;
            this.email = email;
            this.role = role;
        }

        public String getToken() {
            return token;
        }

        public void setToken(String token) {
            this.token = token;
        }

        public String getId() {
            return id;
        }

        public void setId(String id) {
            this.id = id;
        }

        public String getEmail() {
            return email;
        }

        public void setEmail(String email) {
            this.email = email;
        }

        public String getRole() {
            return role;
        }

        public void setRole(String role) {
            this.role = role;
        }
    }
}
